### PR TITLE
Dividing state into parts

### LINKED ISSUE #1237: State sync should not send the entirety of the state in a single message
Currently state sync sends the entirety of the state in a single message.
It should be sending small parts with merkle proofs instead.

Separately, the timeout for the state sync (`STATE_SYNC_TIMEOUT`) most certainly should not be 10 minutes. It should be configurable.

### PR DESCRIPTION
**Pull request by Kouprin**
Fixes #1237.

### CODE REVIEW

File: chain/client/src/client_actor.rs [hunk 1]
```diff
@@ -282,31 +284,52 @@ impl Handler<NetworkClientMessages> for ClientActor {
                     // NetworkClientResponses::Ban { ban_reason: ReasonForBan::BadBlockApproval }
                 }
             }
-            NetworkClientMessages::StateRequest(shard_id, hash) => {
-                if let Ok(shard_state) = self.client.chain.get_state_for_shard(shard_id, hash) {
+            NetworkClientMessages::StateRequest(shard_id, hash, need_header, parts_range) => {
```
Comment (SkidanovAlex) [inline chain/client/src/client_actor.rs]: parts_ranges

File: chain/client/src/client_actor.rs [hunk 2]
```diff
@@ -282,31 +284,52 @@ impl Handler<NetworkClientMessages> for ClientActor {
...
             }
             NetworkClientMessages::StateResponse(StateResponseInfo {
                 shard_id,
                 hash,
                 shard_state,
             }) => {
                 // Populate the hashmaps with shard statuses that might be interested in this state
-                //     (naturally, the plural of statuses is statuseses)
-                let mut shard_statuseses = vec![];
+                let mut shard_to_downloads_vec = vec![];
```
Comment (SkidanovAlex) [inline chain/client/src/client_actor.rs]: shards_to_download

File: chain/client/src/sync.rs [hunk 3]
```diff
@@ -401,52 +402,317 @@ pub enum StateSyncResult {
...
+        tracking_shards: Vec<ShardId>,
+        now: DateTime<Utc>,
+    ) -> Result<(bool, bool), near_chain::Error> {
+        let mut all_done = false;
+        let mut update_sync_status = false;
+
+        for shard_id in tracking_shards {
+            let mut download_timeout = false;
+            let mut need_shard = false;
+            let mut shard_sync_download = &ShardSyncDownload {
```
Comment (SkidanovAlex) [inline chain/client/src/sync.rs]:
```
        let shard_sync_download = 
            if new_shard_sync.contains_key(&shard_id) {
                &new_shard_sync[&shard_id]
            } else {
                need_shard = true;
                &ShardSyncDownload {
                    downloads: vec![
                        DownloadStatus {
                            part_id: 0,
                            start_time: now,
                            prev_update_time: now,
                            run_me: true,
                            error: false,
                            done: false,
                        };
                        1
                    ],
                    status: ShardSyncStatus::StateDownloadHeader,
                }
            };
```

File: chain/client/src/client_actor.rs (hunk 2 above)
Comment (Kouprin) [reply chain/client/src/client_actor.rs]: ++

**Comment by evgenykuzyakov** [comment]
Why ApplyResult contains `num_parts`. How should Runtime even be aware of parts?

**Comment by Kouprin** [comment]
Runtime only decides how many parts will be in the State.

**Comment by Kouprin** [comment]
Please also check https://github.com/nearprotocol/nearcore/compare/staging...T1237?expand=1
`num_parts` is going to be hidden in StateRoot

**Comment by nearmax** [comment]
@Kouprin I talked to @SkidanovAlex  and @mikhailOK and still don't understand why we had to make changes to the runtime. Let's move our discussion here: https://github.com/nearprotocol/nearcore/issues/1447

**Pull request merged**